feat: integrate react SDK for rendering react templates (#453)
asyncapi/generator: docs/authoring.md

1. A template is a directory in your file system.
1. The template can have own dependencies. Just create `package.json` for the template. The generator makes sure to trigger the installation of dependencies.
1. Templates can be configured. Configuration must be stored in the `package.json` file under custom `generator` property. [Read more about the configuration file](#configuration-file).

> 1. The template engine is either [Nunjucks](#nunjucks)(default) or [React](#react). This can be controlled with the `renderer` configuration.

Reviewer: ```suggestion
1. The template engine can be either [React](#react) (recommended) or [Nunjucks](#nunjucks) (default). This can be controlled with the `renderer` property in the [template configuration](#template-configuration).
```

> 1. Templates may contain `hooks` that are functions invoked during specific moment of the generation. In the template, they must be stored in the `.hooks` directory under the template directory. They can also be stored in other modules and external libraries and configured inside the template [Read more about hooks](#hooks).

Reviewer 2: ```suggestion
1. Templates may contain `hooks` that are functions invoked during specific moment of the generation. In the template, they must be stored in the `hooks` directory under the template directory. They can also be stored in other modules and external libraries and configured inside the template [Read more about hooks](#hooks).
```

};
```

## Configuration File

The `generator` property from `package.json` file must contain a JSON object that may have the following information:

|Name|Type|Description|
|---|---|---|

> |`renderer`| String | Supports either `react` or `nunjucks`(default).

Reviewer: ```suggestion
|`renderer`| String | Its value can be either `react` or `nunjucks` (default).
```

|Name|Description|
|---|---|
|`server`| It is used to let the template know which server from the AsyncAPI specification file you want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. Use this parameter in case your template relies on users' information about what server from the specification file they want to use during generation. You also need this parameter if you want to use `server.protocol` notation within `conditionalFiles` configuration option. Once you decide to specify this parameter for your template, it is recommended you make it a mandatory parameter otherwise feature like `conditionalFiles` is not going to work if your users do not use this parameter obligatory.

## React

> [React](https://reactjs.org) is the render engine which we strongly suggest you should use for any new templates. The only reason it is not the default render engine is to stay backward compatibility.

Reviewer: ```suggestion
[React](https://reactjs.org) is the render engine that we strongly suggest you should use for any new templates. The only reason it is not the default render engine is to stay backward compatible.
```

* It enables the possiblity of [debugging](#debugging-react-template) your template (this is not possible with Nunjucks).

> * It provides better error stack traces once something is not working.

Reviewer: ```suggestion
* It provides better error stack traces.
```

I think it's redundant to say "once something is not working".

* Provides better support for separating code into more manageable chunks/components.

> * The readability of the template are much better then Nunjucks syntax.

Reviewer: ```suggestion
* The readability of the template is much better compared to Nunjucks syntax.
```

* Better tool support for development.

> When writing react templates you decide whether to use CommonJS, ES5 or ES6 modules since everything is bundles together before the rendering process takes over. We use our own React renderer which can be found in the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk). 

Reviewer: ```suggestion
When writing React templates you decide whether to use CommonJS, ES5, or ES6 modules since everything is bundled together before the rendering process takes over. We use our own React renderer which can be found in the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk). 
```

There you can find information about how the renderer works or how we transpile your template files.

> Your React template always require 3 dependencies `react`, `source-map-support` and `@asyncapi/generator-react-sdk`. The `react` dependency is required to declare React components inside the template. `source-map-support` ensure that the transpiled files are mapped to the actual template files, this is only to enable correct error stack trace. `@asyncapi/generator-react-sdk` is required to access the `File` component which is required as a root component for a file to be rendered. Furthermore it provides some common components to make your development easier, like `Text` or `Indent`.

Reviewer: ```suggestion
Your React template always requires 3 dependencies `react`, `source-map-support`, and `@asyncapi/generator-react-sdk`. The `react` dependency is required to declare React components inside the template. `source-map-support` ensures that the transpiled files are mapped to the actual template files, this is only to enable correct error stack trace. `@asyncapi/generator-react-sdk` is required to access the `File` component which is required as a root component for a file to be rendered. Furthermore, it provides some common components to make your development easier, like `Text` or `Indent`.
```

> The `react` dependency is required to declare React components inside the template.

Is it? Can't we inject it somehow with Rollup? For instance, when using Next.js it's not necessary. I wonder how they do it.

> `source-map-support` ensures that the transpiled files are mapped to the actual template files, this is only to enable correct error stack trace.

Can't we just include it as part of the `generator-react-sdk`? Just so people don't have to do it.

Author: 🤔 It would be optimal, I'll try and see what I can do.

Author: I have added all the dependencies to the `generator-react-sdk` library. Everything works fine when its used as a normal dependency. However because we in development use a local dependency we are required to add these dependencies our self because npm installs them locally instead of "Merging" them with the template dependencies. 

Not sure how we should communicate this... Maybe I'll add it to the SDK readme? 🤔

Reviewer: Not sure I follow 😅 What do you mean by "merging"? Why is it not working in development?

Author: At the moment in the templates we referer to the SDK with the following: `"@asyncapi/generator-react-sdk": "file:../generator-react-sdk"` i.e. a local dependency. However the way NPM treats this dependency is different from `"@asyncapi/generator-react-sdk": "0.1.0"`.

With a local dependency it installs IT's node_modules inside `../generator-react-sdk/node_modules` i.e. those dependencies are not installed inside `template/node_modules` as it normally would. This means that when the SDK has `source-map-support` dependency the template does not have access to this dependency if we use local dependencies.

Did that make sence in any way? 😅

Reviewer: Oh! I see what you mean. Why is the `generator-react-sdk` a local dependency? Isn't it going to be another published package? I mean, can't we just replace `file:../generator-react-sdk` with `"@asyncapi/generator-react-sdk": "0.1.0"` and it would work?

If you want to work with a local dependency yourself while developing the `generator-react-sdk` package you can use [npm link](https://docs.npmjs.com/cli/v6/commands/npm-link).

Author: Yes thats what we should have done but didnt :D So it should fix itself :)

> Lets consider a basic react template as the one below called `MyTemplate.js`:

Reviewer: ```suggestion
Let's consider a basic React template as the one below called `MyTemplate.js`:
```

```js
import React from "react";
import { File, Text } from "@asyncapi/generator-react-sdk";

> export default function({ asyncapi, params, originalAsyncAPI}) {

Reviewer: > export default function({ asyncapi, params, originalAsyncAPI}) {

I wonder if exporting anonymous functions is the recommended pattern with React. Also, wouldn't it make things harder to debug?

Author: Good question, @magicmatatjahu I think we need your input here 😄

Reviewer 2: At one time, default exports were recommended by the community later that changed 😄 

Return `default` function is crucial for DX and in which way works at this moment the Generator. Why? Because by default export you don't have to name function by `SchemaTemplate` or another predefined name for Generator. We just import the file and then perform some actions to render the template to file. If we enable export by names, then we have a issue: which function we should process? Let's check the example:

```js
export function SomeTemplate({}) {
  // ...implementation
}

export function SomeTemplate2({}) {
  // ...implementation
}
```

which we should choose? As I remember, we didn't discuss on the possibility of rendering several different templates in one file, the difference may be the rendering of an array of templates, which differs only on input data (yes, I am a hypocrite hah 😄).

Reviewer: I don't mean removing `default` but instead adding a function name:

```suggestion
export default function Example({ asyncapi, params, originalAsyncAPI}) {
```

Reviewer 2: Ahhh, ok. Mea culpa. It's should recommended, so we should change the example, but it's not mandatory.

Author: @magicmatatjahu can we mark this as resolved?

  return (

>     <File name={'asyncapi.md'}>

Reviewer: ```suggestion
    <File name="asyncapi.md">
```

      <Text>Some text that should render as is</Text>
    </File>
  );
}
```
The exported default function returns a `File` component as a root component which the [Generator](https://github.com/asyncapi/generator) uses to figure out what file should be generated. In our case we overwrite the default functionality of saving the file as `MyTemplate.js` but instead use the filename `asyncapi.md`. It is then specified that we should render `Some text that should render as is\n` within that file. Notice the `\n` character at the end, this is something that is automatically added after each children in the `File` and `Indent` component. 

For further information about components, props etc. see the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk)

### Common assumptions

1. All files within the template directory are rendered if their root component is of type `File` and the path is not part of the `nonRenderableFiles` configuration.

> 1. New lines are automatically added for each child in `File` and `Indent` component.

Reviewer 2: `New lines are automatically added for each child in `File` and `Indent` component.` is wrong.

New lines are only added in `Text` component. File is w wrapper and Indent adds only indentations (without new lines) :) https://github.com/asyncapi/generator-react-sdk/blob/1e897fc2d2d6b42fb2513f067d88bb5e9b8f2c88/src/utils/withIndendation.ts

### Debugging React template
With React, it enables you to debug your templates. Below is a section of common IDE's and how to enable debugging. 

#### VSCode

> For Visual Studio Code we have created a boiler plate [launch configuration](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations) to enable debug in your template. Add the following launch configuration:

Reviewer: ```suggestion
For Visual Studio Code, we have created a boilerplate [launch configuration](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations) to enable debug in your template. Add the following launch configuration:
```

> With react it enables you to debug your templates. Below are a section of common idé's and how to enable debugging. 

Reviewer: ```suggestion
With React, it enables you to debug your templates. Below is a section of common IDE's and how to enable debugging. 
```

> Add the following launch configuration in your workspace:

Reviewer: Where? I'm using VS Code and honestly have no idea what you mean here 😅

Author: Fair point, I'll add some more documentation here 😄

Author: @fmvilas added some more details and link to launch configuration. Do you feel like more details are needed? 🤔

Reviewer: Looks good now! Thanks. Just added a minor syntax suggestion 👍

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug template",
      "timeout": 10000,
      "sourceMaps": true,
      "args": [
        "./asyncapi.yml",
        "./template",
        "--output",
        "./output",
        "--install",
        "--force-write"
      ],
      "program": "ag"
    }
  ]
}
```
Now replace `./asyncapi.yml` with your document of choice. Replace `./template` with the path to your React template. You can now debug your template by adding any breakpoints you want and inspect your code.

## Nunjucks

> [Nunjucks](https://mozilla.github.io/nunjucks) is the default render engine, however we strongly suggest you to consider adapting to the [React](#react) engine.

Reviewer: ```suggestion
[Nunjucks](https://mozilla.github.io/nunjucks) is the default render engine, however, we strongly recommend adopting the [React](#react) engine.
```

Let's consider a basic React template as the one below called `MyTemplate.js`:

```js
import { File, Text } from "@asyncapi/generator-react-sdk";

export default function({ asyncapi, params, originalAsyncAPI }) {
  return (
    <File name="asyncapi.md">
      <Text>Some text that should render as is</Text>
    </File>
  );
}
```

The exported default function returns a `File` component as a root component which the [Generator](https://github.com/asyncapi/generator) uses to figure out what file should be generated. In our case we overwrite the default functionality of saving the file as `MyTemplate.js` but instead use the filename `asyncapi.md`. It is then specified that we should render `Some text that should render as is\n` within that file. Notice the `\n` character at the end, this is something that is automatically added after `Text` component. 

For further information about components, props etc. see the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk)

### Common assumptions

> 1. All files within the template directory are rendered if their root component is of type `File` and the path is not part of the `nonRenderableFiles` configuration.

Author: Yea, much better 😄

1. New lines are automatically added after each `Text` component.

### Debugging React template
With React, it enables you to debug your templates. Below is a section of common IDE's and how to enable debugging. 

> #### VSCode

Author: The thought was that it could be introduced as [suggested](https://github.com/asyncapi/generator/pull/453#discussion_r536196752) by @magicmatatjahu. I am neutral about this, I can understand why it should be changed since we only provide one example.

> #### VSCode

Reviewer 3: exactly, so once we have more, we can introduce structure, now it looks strange unless you have another PR coming already 😄

> #### VSCode

Author: yea I actually agree, I'll change it.

  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug template",
      "timeout": 10000,
      "sourceMaps": true,
      "args": [
        "./asyncapi.yml",
        "./template",
        "--output",
        "./output",
        "--install",
        "--force-write"
      ],
      "program": "ag"
    }
  ]
}
```
Now replace `./asyncapi.yml` with your document of choice. Replace `./template` with the path to your React template. You can now debug your template by adding any breakpoints you want and inspect your code.

## Nunjucks
[Nunjucks](https://mozilla.github.io/nunjucks) is the default render engine, however, we strongly recommend adopting the [React](#react) engine.
### Common assumptions

1. Templates may contain [Nunjucks filters or helper functions](https://mozilla.github.io/nunjucks/templating.html#builtin-filters). [Read more about filters](#filters).
1. Templates may contain `partials` (reusable chunks). They must be stored in the `.partials` directory under the template directory. [Read more about partials](#partials).
1. Templates may contain multiple files. Unless stated otherwise, all files will be rendered.

> 1. The default variables you have access to in any the template file are the following:

Author: I am guessing its only the variables you refer to. Even though its the same variables they are provided through the props, which kinda needed its own description. Unless you think people will understand the difference ? 🤔

> 1. The default variables you have access to in any the template file are the following:

Reviewer 3: for me, the meaning behind this point was to explain to template developer what is the context of the nunjucks template, what data I have access to, that it is the spec before and after parsing + config params.

How it works with react? what do I have access to in react template? in my link, I refer to the object argument you pass there to the function

> 1. The default variables you have access to in any the template file are the following:

Author: That is a very good question.. I actually thought I added that 👻 I guess not 🤷

> 1. The default variables you have access to in any the template file are the following:

Author: @derberg I have added it to the [common assumptions](https://github.com/asyncapi/generator/blob/be7dc5b83da90986ac7215a206f46341ddb7a2ac/docs/authoring.md#common-assumptions-1), what you think?

> 1. The default variables you have access to in any the template file are the following:

Reviewer 3: @jonaslagoni isn't it duplicated content? I mean shouldn't we just put it where it was, at the top of the document, if it is for any render engine?

> 1. The default variables you have access to in any the template file are the following:

Author: From my POV it is different ways it gets provided, i.e. nunjucks provide it globally, and React provide it though props. But yea the props themself are the same, the context in which they are provided are different and has to be stated. Not sure what makes most sense when seen from the outside? 🤷

> 1. The default variables you have access to in any the template file are the following:

Reviewer 3: @jonaslagoni oh yes, sorry, you are right, technically it is a difference as you stated so makes sense to describe it the way you did 👍🏼

|Name|Description|
|---|---|
|`server`| It is used to let the template know which server from the AsyncAPI specification file you want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. Use this parameter in case your template relies on users' information about what server from the specification file they want to use during generation. You also need this parameter if you want to use `server.protocol` notation within `conditionalFiles` configuration option. Once you decide to specify this parameter for your template, it is recommended you make it a mandatory parameter otherwise feature like `conditionalFiles` is not going to work if your users do not use this parameter obligatory.

## React

[React](https://reactjs.org) is the render engine that we strongly suggest you should use for any new templates. The only reason it is not the default render engine is to stay backward compatible.

> 

Reviewer 3: please add also an important point about components reusability that we could not achieve with nunjucks (like share template file across code templates for the same language)

Author: @derberg but thats also possible with nunjucks right? You just have to point to the direct source in node_modules when installed as a dependency 🤔 ?

Reviewer 3: oh, you mean like add a template to dependency and then point to it directly, like when importing macro? Technically it works, but it is error-prone, you kinda hardcode yourself to a template file structure, we would never go this way

Reviewer 3: also, components testability improves how we can test template's output

Author: Does not have to be a template dependency, could be a library containing nunjucks partials in some directory. Not quite sure how the generator behaves in regards to outside partials. however do you still think we should add the reason since it is already stated right below in the list 🤔 ?

Author: > also, components testability improves how we can test template's output

True, this is missing.

Reviewer 3: I think so, because anything else with nunjucks would always be dependent on files structure. So if in a possible library of reusable macros, you would change folder structure, that would be a breaking change, something you do not want and it is easy to forget as it doesn't feel natural. This is in my opinion why we never even considered doing this ourselves and promoting such support.

Because we use React, we do not have such a problem.

Author: @derberg added the part about testability also the part about reusability is also stated here -> https://github.com/asyncapi/generator/blob/775d1ea6032fc9e403bb23afd100f317771af9f3/docs/authoring.md#react

|Name|Description|
|---|---|
|`server`| It is used to let the template know which server from the AsyncAPI specification file you want to use. In some cases, this may be required. For instance, when generating code that connects to a specific server. Use this parameter in case your template relies on users' information about what server from the specification file they want to use during generation. You also need this parameter if you want to use `server.protocol` notation within `conditionalFiles` configuration option. Once you decide to specify this parameter for your template, it is recommended you make it a mandatory parameter otherwise feature like `conditionalFiles` is not going to work if your users do not use this parameter obligatory.

## React

> [React](https://reactjs.org) is the newest render engine which we strongly suggest you should use for any new templates. It is only the default render engine to enable backward compatibility.

Author: Yea sounds better. Last section was to describe why it was not the default render engine but yea was missing some words, changed it 😄

* It enables the possiblity of [debugging](#debugging-react-template) your template (this is not possible with Nunjucks).
* It provides better stack traces once something is not working.
* Provides better support for separating code into more manageable chunks.

> * The readability of the template are much better then Nunjucks syntax.

Author: I think we should keep it because it outlines the issues that is inherit by Nunjucks that they will encounter if they go for that instead. People can make their own decision, but I think it is a good idea to have it 😄

* Better tool support for development.

When writing react templates you decide whether to use CommonJS, ES5 or ES6 modules since everything is bundles together before the rendering process takes over. We use our own React renderer which can be found in the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk). 
There you can find information about how the renderer works or how we transpile your template files.

> Your React template always require 2 dependencies `react` and `@asyncapi/generator-react-sdk`. The `react` dependency is pretty self explanitory, it allows us to render the template. `@asyncapi/generator-react-sdk` is required since react files does not export default function where the root component is not of type `File` will not be rendered, furthermore it provide some support components to easy your development. 

Reviewer 2: ```md
...The `react` dependency is pretty self explanitory, it allows us to render the template...
```

The react is required because before transpilation you have `<Component {...props} >` but after transpilation you have `React.createElement(Component, props)` so without `import React from 'react'` then your code is broken (React will be undefined). It's a magic but it's necessary :) Also react doesn't render our template - it does by our custom renderer in `@asyncapi/generator-react-sdk` package by exported `render` method.

I would rather to change whole paragraph to something like:

```md
Your React template always require 2 dependencies `react` and `@asyncapi/generator-react-sdk`. The `react` dependency is required to declare React components inside template. `@asyncapi/generator-react-sdk` is required to import and use as a root component of your template the `File` component - it provides required metadata and without it your template won't be rendered. Furthermore it provides some common components to make your development easier, like `Text` or `Indent`
```

Author: Yea you are right, abstracted away from that a bit. Try see the updated readme 😄

Lets consider a basic react template as the one below called `MyTemplate.js`:
```js
import React from "react";
import { File, Text } from "@asyncapi/generator-react-sdk";
export default function({ asyncapi, params, originalAsyncAPI}) {
  return (
    <File name={'asyncapi.md'}>
      <Text>Some text that should render as is</Text>
    </File>
  );
}

> ```

Reviewer 2: Maybe we should change this example to this https://github.com/asyncapi/generator-react-sdk/blob/4ed0acffc9c674cdf2f3a4d67feac4a5fb53ca3b/README.md#example to show how to using the custom components? 😄

Reviewer 2: I mean change the content of `File` component :)

Author: hmm, I wanted to keep it as simple as possible, also why in the end I refer to the SDK for further information about props, components etc 🤔

> The react component returns a `File` component as root component which the React renderer uses to figure out what what file should be generated. In our case we overwrite the default functionality of saving the file as `MyTemplate.js` but instead use the filename `asyncapi.md`. It is then specified that we should render `Some text that should render as is` within that file.

Reviewer 2: ```md
The exported by default function returns a `File` component as a root component which the [Generator](https://github.com/asyncapi/generator) uses to figure out what file should be generated. In our case we overwrite the default functionality of saving the file as `MyTemplate.js` but instead use the filename `asyncapi.md`. It is then specified that we should render `Some text that should render as is\n` within that file.
```

New line is very important 😄 and also I changed that Generator figure out metadata not renderer. Renderer only travserses components hierarchy and changes components to string.

Author: good point, added.

For further information about components, props etc. see the [Generator React SDK](https://github.com/asyncapi/generator-react-sdk)

> ### Debugging React template

Author: Hmm, yea not sure, I like when libraries make it as easy as possible. This ensures that they dont need to figure out how to setup their own actions, but I see your point..

> ### Debugging React template

Reviewer 2: Ok, I understand, but maybe we should add subheaders for appropriate IDEs? :) `### Debugging React template` then `#### VSC` etc.. What do you think?

  );
}
```

> The exported default function returns a `File` component as a root component which the [Generator](https://github.com/asyncapi/generator) uses to figure out what file should be generated. In our case we overwrite the default functionality of saving the file as `MyTemplate.js` but instead use the filename `asyncapi.md`. It is then specified that we should render `Some text that should render as is\n` within that file. Notice the `\n` character at the end, this is something that is automatically added after each children in the `File` and `Indent` component. 

Reviewer 2: ```md
Notice the `\n` character at the end, this is something that is automatically added after each children in the `File` and `Indent` component. 
```

It's added by each `Text` component, not after each children 😄 https://github.com/asyncapi/generator-react-sdk/blob/1e897fc2d2d6b42fb2513f067d88bb5e9b8f2c88/src/components/Text.tsx#L50
